File: 2025/py/day01.py
import os
import time


# with open(os.path.join(os.path.abspath(os.path.dirname(__file__)), "../input/sample.txt"), "r") as f:
with open(os.path.join(os.path.abspath(os.path.dirname(__file__)), "../input/day01.txt"), "r") as f:
    rots = f.readlines()

# p1

start = time.perf_counter()
zeroes = 0
point = 50
for rot in rots:
    diff = 100 - int(rot[1:]) if rot[0] == 'L' else int(rot[1:])
    point = ((point + diff) % 100)
    if point == 0:
        zeroes += 1
end = time.perf_counter()

print(f'P1: {zeroes} (time: {end - start:.6f} s)')

# p2

start = time.perf_counter()
zeroes = 0
point = 50
for rot in rots:
    diff = int(rot[1:])
    if rot[0] == 'L':
        p = point - diff
        if p == 0 and point > 0:
            zeroes += 1
        elif p < 0:
            if point > 0:
                zeroes += 1
            zeroes += abs(p) // 100
            p = (p + 100) % 100
        point = p
    else:
        point += diff
        zeroes += point // 100
        point %= 100
end = time.perf_counter()

print(f'P2: {zeroes} (time: {end - start:.6f} s)')

File: 2025/bin/day01.ml
open StdLabels

(* let input = Aoc.read_sample_lines () *)
let input = Aoc.read_input_lines 1

let parse_rot rot = String.sub rot ~pos:1 ~len:(String.length rot - 1) |> int_of_string

let rotate pos rot =
  pos
  +
  match rot.[0] with
  | 'L' -> 100 - parse_rot rot
  | _ -> parse_rot rot

let rec solve rots pos =
  (if pos = 0 then 1 else 0)
  +
  match rots with
  | [] -> 0
  | rot :: rest -> solve rest (rotate pos rot mod 100)

(* part 1 *)
let () =
  let t0 = Unix.gettimeofday () in
  let ans = solve input 50 in
  let t1 = Unix.gettimeofday () in
  Printf.printf "P1: %d (%f s)\n" ans (t1 -. t0)

let pos_mod a b = ((a mod b) + b) mod b

let rec solve rots pos =
  match rots with
  | [] -> 0
  | rot :: rest -> (
      match rot.[0] with
      | 'L' ->
          let zeroes, p =
            match (pos - parse_rot rot, pos) with
            | 0, prev when prev > 0 -> (1, 0)
            | p, prev when p < 0 -> (((abs p / 100) + if prev > 0 then 1 else 0), pos_mod p 100)
            | p, _ -> (0, p)
          in
          zeroes + solve rest p
      | _ ->
          let p = pos + parse_rot rot in
          (p / 100) + solve rest (p mod 100))

(* part 2 *)
let () =
  let t0 = Unix.gettimeofday () in
  let ans = solve input 50 in
  let t1 = Unix.gettimeofday () in
  Printf.printf "P2: %d (%f s)\n" ans (t1 -. t0)
